larreco/QuadVtx/QuadVtx_module.cc

```diff
@@ -0,0 +1,522 @@
+struct Pt2D
+{
+  Pt2D(double _x, double _z, int _view, double _energy) : x(_x), z(_z), view(_view), energy(_energy) {}
+
+  bool operator<(const Pt2D& p) const {return z < p.z;}
+
+  double x, z;
+  int view;
+  double energy;
+};
+
+
+// ---------------------------------------------------------------------------
+struct Line2D
+{
+  Line2D(const Pt2D& a, const Pt2D& b)
+    : m((b.x-a.x)/(b.z-a.z)), c(b.x-m*b.z),// w(a.energy * b.energy),
+      minz(std::min(a.z, b.z)), maxz(std::max(a.z, b.z))
+  {
+  }
+
+  inline bool operator<(const Line2D& l) const {return m < l.m;}
+
+  float m, c, /*w,*/ minz, maxz;
+};
+
+// ---------------------------------------------------------------------------
+class QuadVtx: public art::EDProducer
+{
+public:
+  explicit QuadVtx(const fhicl::ParameterSet& pset);
+
+  void beginJob() override;
+  void produce(art::Event& evt) override;
+  void endJob() override;
+
+protected:
+  bool FindVtx(const std::vector<recob::Hit>& hits,
+               TVector3& vtx,
+               art::TFileDirectory* evt_dir) const;
+
+  std::string fHitLabel;
+
+  bool fSavePlots;
+
+  const detinfo::DetectorProperties* detprop;
+  const geo::GeometryCore* geom;
+};
+
+DEFINE_ART_MODULE(QuadVtx)
+
+// ---------------------------------------------------------------------------
+QuadVtx::QuadVtx(const fhicl::ParameterSet& pset) :
+  EDProducer(pset),
+  fHitLabel(pset.get<std::string>("HitLabel")),
+  fSavePlots(pset.get<bool>("SavePlots"))
+{
+  produces<std::vector<recob::Vertex>>();
+}
+
+// ---------------------------------------------------------------------------
+void QuadVtx::beginJob()
+{
+  detprop = art::ServiceHandle<const detinfo::DetectorPropertiesService>()->provider();
+  geom = art::ServiceHandle<const geo::Geometry>()->provider();
+}
+
+// ---------------------------------------------------------------------------
+void QuadVtx::endJob()
+{
+}
+
+// ---------------------------------------------------------------------------
+// x = m*z+c. z1 and z2 are the two intercepts in case of returning true
+bool IntersectsCircle(float m, float c,
+                      float z0, float x0,
+                      float R,
+                      float& z1, float& z2)
+{
+  // Change to the frame where (z0, x0) = (0, 0)
+  c += m*z0 - x0;
+
+  // z^2 + (m*z+c)^2 = R^2
+  const float A = 1+sqr(m);
+  const float B = 2*m*c;
+  const float C = sqr(c)-sqr(R);
+
+  double desc = sqr(B)-4*A*C;
+
+  if(desc < 0) return false;
+
+  desc = sqrt(desc);
+
+  z1 = (-B-desc)/(2*A);
+  z2 = (-B+desc)/(2*A);
+
+  // Back to the original frame
+  z1 += z0;
+  z2 += z0;
+
+  return true;
+}
+
+// ---------------------------------------------------------------------------
+void LinesFromPoints(const std::vector<Pt2D>& pts,
+                     std::vector<Line2D>& lines,
+                     float z0 = 0, float x0 = 0, float R = -1)
+{
+  const size_t kMaxLines = 10*1000*1000; // This is 150MB of lines...
```

Comment:
Suggest `constexpr size_t` instead of `const size_t`.

Reply:
Done, though I suspect it amounts to exactly the same thing for a simple case like this.